src/MSBuildTaskHost/Immutable/ImmutableDictionary.cs

```diff
@@ -0,0 +1,247 @@
+        {
+            return new ImmutableDictionary<K, V>(dictionary);
+        }
+    }
+
+    static class ImmutableDictionary
+    {
+        internal static ImmutableDictionary<K, V> Create<K, V>(IEqualityComparer<K> comparer)
+        {
+            return new ImmutableDictionary<K, V>(comparer);
+        }
+    }
+
+    /// <summary>
+    /// Inefficient ImmutableDictionary implementation: keep a mutable dictionary and wrap all operations.
+    /// </summary>
+    /// <typeparam name="K"></typeparam>
+    /// <typeparam name="V"></typeparam>
+    sealed class ImmutableDictionary<K, V> : IDictionary<K, V>, IDictionary
+    {
+        /// <summary>
+        /// The underlying dictionary.
+        /// </summary>
+        private Dictionary<K, V> _backing;
+
+        //
+        // READ-ONLY OPERATIONS
+        //
+
+        public ICollection<K> Keys => _backing.Keys;
+        public ICollection<V> Values => _backing.Values;
+
+        ICollection IDictionary.Keys => _backing.Keys;
+        ICollection IDictionary.Values => _backing.Values;
+
+        public int Count => _backing.Count;
+
+        public V this[K key] => _backing[key];
+
+        public bool IsReadOnly => true;
+        public bool IsFixedSize => true;
+        public bool IsSynchronized => true;
+
+        public object SyncRoot => this;
+
+        public bool TryGetValue(K key, out V value)
+        {
+            return _backing.TryGetValue(key, out value);
+        }
+
+        public bool Contains(KeyValuePair<K, V> item)
+        {
+            return _backing.Contains(item);
+        }
+
+        bool IDictionary.Contains(object key)
+        {
+            return ((IDictionary)_backing).Contains(key);
+        }
+
+        public bool ContainsKey(K key)
+        {
+            return _backing.ContainsKey(key);
+        }
+
+        public IEnumerator<KeyValuePair<K, V>> GetEnumerator()
+        {
+            return _backing.GetEnumerator();
+        }
+
+        IDictionaryEnumerator IDictionary.GetEnumerator()
+        {
+            return _backing.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return _backing.GetEnumerator();
+        }
+
+        void ICollection<KeyValuePair<K, V>>.CopyTo(KeyValuePair<K, V>[] array, int arrayIndex)
+        {
+            foreach (var item in this)
+            {
+                array[arrayIndex++] = item;
+            }
+        }
+
+        void ICollection.CopyTo(Array array, int arrayIndex)
+        {
+            foreach (var item in this)
+            {
+                array.SetValue(new DictionaryEntry(item.Key, item.Value), arrayIndex++);
+            }
+        }
+
+        //
+        // WRITE OPERATIONS
+        //
+
+        internal ImmutableDictionary<K, V> SetItem(K key, V value)
+        {
+            var clone = new ImmutableDictionary<K, V>(_backing);
+            clone._backing.Add(key, value);
```

Comment:
We could make the check in `Remove` but for `SetItem` we would have to have the concept of value comparer. From `ImmutableDictionary` docs:

_If the specified key/value pair already exists in the dictionary, this method returns the existing instance of the dictionary. If the key already exists but has a different value, this method returns a new instance of the dictionary with the overwritten value._

Also, fixed to call the indexer instead of `Add` on the backing dictionary to comply with the second sentence above.

Reply:
I think that even if the values are identical for all normal purposes, ImmutableDictionary still considers them different if they're different instances, so we could just use `!=`.

The way it currently is, if you rely on having the exact same object returned from SetItem before and after you try to add the same key with the same value, it'll work for the real implementation and fail here. That's such a trivial case that I don't think it's worth fixing, but it's worth bringing up. In contrast, changing from Add to the indexer changes whether it throws an error, which was worth it.